collectors/osd.go

```diff
@@ -752,33 +955,71 @@ func (o *OSDCollector) collectOSDScrubState(ch chan<- prometheus.Metric) error {
 	}
 
 	for i, v := range o.osdScrubCache {
+		lb := o.getOSDLabelFromID(int64(i))
 		ch <- prometheus.MustNewConstMetric(
 			o.ScrubbingStateDesc,
 			prometheus.GaugeValue,
 			float64(v),
-			fmt.Sprintf(osdLabelFormat, i))
+			fmt.Sprintf(osdLabelFormat, i),
+			lb.DeviceClass,
+			lb.Host,
+			lb.Root,
+			lb.Root)
 	}
 
 	return nil
 }
 
 func (o *OSDCollector) collectPGRecoveryState(ch chan<- prometheus.Metric) error {
 	for _, pg := range o.pgDumpBrief.PGStats {
-		if o.initialCollect || strings.Contains(pg.State, "recovering") {
+
+		// We need previous PG state in order to update the metric when a PG has
+		// completed recovery or backfill. Or it could be an empty string if
+		// unknown.
+		prevPGState, prevPGStateFound := o.pgStateCache[pg.PGID]
+		prevNumObjectsRecovered := o.pgObjectsRecoveredCache[pg.PGID]
+		prevBackfillTargets := o.pgBackfillTargetsCache[pg.PGID]
+
+		if !prevPGStateFound || strings.Contains(prevPGState, "recovering") || strings.Contains(pg.State, "recovering") ||
+			strings.Contains(prevPGState, "backfilling") || strings.Contains(pg.State, "backfilling") {
 			query, err := o.performPGQuery(pg.PGID)
 			if err != nil {
 				continue
 			}
 
+			o.pgStateCache[pg.PGID] = pg.State
 			o.pgObjectsRecoveredCache[pg.PGID] = query.Info.Stats.StatSum.NumObjectsRecovered
+			o.pgBackfillTargetsCache[pg.PGID] = query.backfillTargets()
+
+			// There is no previous backfill_targets, and we have just cached
+			// it. Wait for the next time so that we can know the increased
+			// number of objects backfilled for this entire PG and compute the
+			// average increased number of objects backfilled for each OSD in
+			// the backfill_targets.
+			if prevBackfillTargets == nil || len(prevBackfillTargets) == 0 {
```

Comment:
Should we move this check closer to where `prevBackfillTargets := o.pgBackfillTargetsCache[pg.PGID]` is declared?

Reply:
We could, but this was a direct diff/port over from luminous (IE same place it's in the luminous code base ).

Reply:
Cool, works for me!